addons/api/addon/models/role.js

```diff
@@ -37,6 +39,124 @@ export default class RoleModel extends GeneratedRoleModel {
       : null;
   }
 
+  /**
+   * A list of IDs for principals of type `user`.
+   * @type {string[]}
+   */
+  get userIDs() {
+    return this.principals
+      .filter(({ type }) => type === 'user')
+      .map(({ principal_id }) => principal_id);
+  }
+
+  /**
+   * A list of IDs for principals of type `group`.
+   * @type {string[]}
+   */
+  get groupIDs() {
+    return this.principals
+      .filter(({ type }) => type === 'group')
+      .map(({ principal_id }) => principal_id);
+  }
+
+  /**
+   * A list of IDs for principals of type `managed group`.
+   * @type {string[]}
+   */
+  get managedGroupIDs() {
+    return this.principals
+      .filter(({ type }) => type === 'managed group')
+      .map(({ principal_id }) => principal_id);
+  }
+
+  /**
+   * A promise that resolves to an array of User instances.
+   * When calling this getter, be sure to await resolution
+   * before interacting with the results.
+   * @type {Promise[UserModel]}
+   */
+  get users() {
+    const ids = this.userIDs;
+
+    // Role has prinicipal IDs,
+    // return a promise which resolves model instances for those IDs
+    if (ids?.length) {
+      return this.resourceFilterStore
+        .queryBy('user', { id: ids }, { scope_id: 'global', recursive: true })
+        .then((models) => models.map((model) => model));
+    }
+
+    // No principal IDs,
+    // return a promise resolving to an empty array
+    return resolve([]);
+  }
+
+  /**
+   * A promise that resolves to an array of Group instances.
+   * When calling this getter, be sure to await resolution
+   * before interacting with the results.
+   * @type {Promise[GroupModel]}
+   */
+  get groups() {
+    const ids = this.groupIDs;
+
+    // Role has prinicipal IDs,
+    // return a promise which resolves model instances for those IDs
+    if (ids?.length) {
+      return this.resourceFilterStore
+        .queryBy('group', { id: ids }, { scope_id: 'global', recursive: true })
+        .then((models) => models.map((model) => model));
+    }
+
+    // No principal IDs,
+    // return a promise resolving to an empty array
+    return resolve([]);
```

Comment:
When should we still be using `rsvp` vs just normal promises? The [docs](https://api.emberjs.com/ember/4.4/functions/rsvp/resolve) says it just aliases to promises but are we doing some thing with the ember runloop or could this just be a normal promise?

Reply:
It looks like [Ember is indeed integrating the runloop with RSVP](https://github.com/emberjs/ember.js/blob/master/packages/%40ember/-internals/runtime/lib/ext/rsvp.ts).  I haven't experimented with mixing native promises with RSVP.  It should technically be possible as long as RSVP implements the interface.  To be safe and idiomatically Ember, we tend to stick with RSVP though.

Reply:
It does seem like `RSVP` is a product of its times considering it was before ES6 and the widespread adoption of native promises, this [thread](https://discuss.emberjs.com/t/should-we-still-use-rsvp-instead-of-native-promises/19243) and some of the linked threads seem to indicate it should be fine to use native promises now, it just has some effects in testing as the `RSVP` promises automatically wait for the app to "settle".

Reply:
maybe this could be a great tech time topic, given that we use RSVP in many places